packages/block-editor/src/components/global-styles/advanced-panel.js

```diff
@@ -37,40 +38,130 @@ export default function AdvancedPanel( {
 			}
 		}
 	}
-	function handleOnBlur( event ) {
-		if ( ! event?.target?.value ) {
+	function handleOnBlur( newValue ) {
+		if ( ! newValue ) {
 			setCSSError( null );
 			return;
 		}
-
 		const [ transformed ] = transformStyles(
-			[ { css: event.target.value } ],
+			[ { css: newValue } ],
 			'.editor-styles-wrapper'
 		);
-
 		setCSSError(
 			transformed === null
 				? __( 'There is an error with your CSS structure.' )
 				: null
 		);
 	}
 
+	/**
+	 * Ensure the editor has at least min lines of code,
+	 * as the editor will shrink to fit the content.
+	 * @param {string} content The content to ensure min lines for.
+	 * @return {string} The content with at least min lines.
+	 */
+	function ensureMinLines( content ) {
+		const MIN_LINES = 10;
+		const LINE_HEIGHT = 18.2; // Height of one line in the editor
+		const MARGIN = 53.4;
+		let requiredLines = MIN_LINES;
+
+		const lines = content.split( '\n' );
+		const contentLineCount = lines.length;
+
+		const wrapper = document.querySelector(
+			'.edit-site-global-styles-screen-css'
+		);
+		if ( wrapper ) {
+			const wrapperHeight = wrapper.offsetHeight;
+			const editorHeight = wrapperHeight - MARGIN;
+
+			// Calculate the minimum number of lines that should be displayed
+			const calcMinLineCount = Math.ceil( editorHeight / LINE_HEIGHT );
+			requiredLines = Math.max( MIN_LINES, calcMinLineCount );
+
+			// Set the max height of the editor allowing scrolling by `overflow-y: scroll`
+			const editor = document.getElementById( EDITOR_ID );
+			if ( editor ) {
+				editor.style.height = `${ editorHeight }px`;
+			}
+		}
+
+		let result = content;
+		for ( let i = contentLineCount; i < requiredLines; i++ ) {
+			result += '\n';
+		}
+
+		return result;
+	}
+	const editorRef = useRef();
+	useEffect( () => {
+		( async () => {
+			/**
+			 * Lazy load CodeMirror by using Webpack's dynamic import.
+			 * This should be replaced with native dynamic import once it's supported.
+			 * @see https://github.com/WordPress/gutenberg/pull/60155
+			 */
+			const { EditorView, basicSetup } = await import( 'codemirror' );
+			const { indentWithTab } = await import( '@codemirror/commands' );
+			const { keymap } = await import( '@codemirror/view' );
+			const { css } = await import( '@codemirror/lang-css' );
```

Comment:
Importing these individual packages will create four webpack chunks, and the loading will take more time than necessary, because the next load starts only after the previous one has finished.

Better to move the `EditorView` code into a separate module and load it all at once:
```js
const editorRef = useRef();
useEffect( () => {
  import( './code-mirror-view' ).then( ( { default: createEditor } ) => {
    createEditor( {
      parent: editorRef.current,
      mode: 'css', // or 'html'
      onChange: handleOnChange,
      onBlur: handleOnBlur,
    } );
  } );
}, [] );
```

Then there will be just one `code-mirror-view` chunk that's loaded on demand. There's a bit of trouble with the `lang-css` and `lang-html` modules. Do we want to bundle them both in a single chunk? That depends on how big they are.

Or the dynamic module can contain the React component that shows the editor: the `<div>` together with the `useRef` and the `useEffect` (this time the effect doesn't do dynamic `import`). Then it can be used like:
```js
const CodeEditor = React.lazy( () => import( './code-mirror-view' ) );

<Suspense fallback={ <Placeholder /> }>
  <CodeEditor id className aria-describedBy value onChange onBlur />
</Suspense>
```

It would be also a good idea to not load the CodeMirror editor not when the `AdvancedPanel` is rendered, that will load it too often. Load it only after clicking on the CSS field, only after the user shows a clear intent to edit.

Reply:
Thank you for your review, @jsnajdr! 😄 

>and the loading will take more time than necessary, because the next load starts only after the previous one has finished

Not exactly, in terms of **`import`**. Fundamentally, based on my understanding of [how modules are waited for and executed](https://github.com/tc39/proposal-top-level-await#semantics-as-desugaring), the following code

```php
import { a } from './a.mjs';
import { b } from './b.mjs';
import { c } from './c.mjs';

console.log(a, b, c);
```

would be roughly equivalent to the following: 

```php
import { promise as aPromise, a } from './a.mjs';
import { promise as bPromise, b } from './b.mjs';
import { promise as cPromise, c } from './c.mjs';

export const promise = Promise.all([aPromise, bPromise, cPromise]).then(() => {
	console.log(a, b, c);
});
```

At least, Webpack behaves in this manner, even for **`import()`**. Here’s a screenshot showing that the requests for chunks run **in parallel** when using the original approach.

<img width="1295" alt="Screenshot 2024-03-28 at 13 47 15" src="https://github.com/WordPress/gutenberg/assets/5287479/8763f12e-bf4e-467a-ad52-82d8cd95b8b4">

**However**, this does not seem to be stated explicitly in [the](https://github.com/tc39/proposal-dynamic-import) [spec](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-import-calls) of `import()`, and my tests confirmed that native dynamic imports execute requests **sequentially** both at the top level and within function scopes. I’ll address this later in this comment. 

>Importing these individual packages will create four webpack chunks

This also applies even if we:

- consolidate everything into one React component,
- AND use `React.lazy()`
- AND use static `import {} from 'codemirror'`

because we are [splitting the vendor's chunk](https://github.com/WordPress/gutenberg/blob/59a81b5666a1ee6bb69dbe115c177c1c3bc0d0b9/packages/scripts/config/webpack.config.js#L138). 

This approach does not create a single chunk, even with all logic extracted into a component and using static import (although we can tweak chunks further by using the [optimization.splitChunks](https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks) option).

Thus, the requests for the “one-component approach” looked like this, showing no difference from the original approach:

<img width="1294" alt="Screenshot 2024-03-28 at 13 55 10" src="https://github.com/WordPress/gutenberg/assets/5287479/eae4a729-ad14-4206-b393-e4d3220071f4">

**Given the above, I propose the following;**

- For explicit parallelization, use **`Promise.all([import(), import(), …])`** rather than fetching a single larger chunk at once. This ensures that requests run in parallel, even after transitioning to native dynamic import.
- Extract “EditorView”: I favor the one-component approach for its cohesiveness (rather than for performance reasons), so I'll encapsulate the logic into a component for reuse. 🙂

For now, parallel chunk requests seem to be more efficient.

>Then there will be just one `code-mirror-view` chunk that's loaded on demand. There's a bit of trouble with the `lang-css` and `lang-html` modules. Do we want to bundle them both in a single chunk? That depends on how big they are.

I don’t think it’s a good idea to load all **`lang-*`** modules in this component because we may want to support additional language modes in the future. (Even with the “parallel requests” strategy, this could potentially max out the browser's limit on simultaneous requests.)

>It would be also a good idea to not load the CodeMirror editor not when the AdvancedPanel is rendered, that will load it too often. Load it only after clicking on the CSS field, only after the user shows a clear intent to edit.

On a slow 3G network, this delay could be around 8 seconds. Making users wait that long for syntax highlighting isn't ideal, IMO.

I'd consider more eager loading strategies (if that doesn't interfere with user interaction). For instance, for Additional CSS, we could load the chunks when the three-dots menu is opened, or for a Custom HTML block, load them when the block appears in the search results. Alternatively, **`prefetch`** might be a viable approach. WDYT?

Once the chunks are loaded, subsequent network requests are unnecessary, even if users access Additional CSS multiple times. Thus, “loading them too often” shouldn’t pose an issue, in my opinion. We just need to be careful about how it affects users’ actual interactions/experiences.